[coreclr] Native AOT runtime and functional tests on apple mobile report undefined symbols for architecture arm64

## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=367843&view=logs&jobId=1a7f2d0f-403f-5f1a-afbb-20466d7ac54e
Leg Name: ios-arm64 Release AllSubsets_NativeAOT_RuntimeTests / ios-arm64 Release AllSubsets_NativeAOT

<!-- Error message template  -->
## Error Message
<!-- Fill for repository issues. For infrastructure issues the engineering services (@dotnet/dnceng) is going to fill it. -->
```json 
{ 
    "ErrorMessage": "Undefined symbols for architecture arm64",
    "BuildRetry": false,
    "ErrorPattern": "",
    "ExcludeConsoleLog": false
} 
```
The regression may have been introduced in https://github.com/dotnet/runtime/pull/89869. /cc: @steveisok @akoeplinger 

Error log:
```
/tmp/helix/working/B02E09DD/w/AD93097C/e/apple_build/AppBundle/iOS.Device.Aot.Test/iOS.Device.Aot.Test.build/Release-iphoneos/iOS.Device.Aot.Test.build/Objects-normal/arm64/iOS.Device.Aot.Test_dependency_info.dat -o /tmp/helix/working/B02E09DD/w/AD93097C/e/apple_build/AppBundle/iOS.Device.Aot.Test/Release-iphoneos/iOS.Device.Aot.Test.app/iOS.Device.Aot.Test [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018: Undefined symbols for architecture arm64: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_AppendExceptionStackFrame", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_FailFast", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_GetRuntimeException", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_GetSystemArrayEEType", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_IDynamicCastableGetInterfaceImplementation", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_IDynamicCastableIsInterfaceImplemented", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_InitializeModules", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       InitializeRuntime() in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_ObjectiveCMarshalGetIsTrackedReferenceCallback", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_ObjectiveCMarshalGetOnEnteredFinalizerQueueCallback", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_ObjectiveCMarshalGetUnhandledExceptionPropagationHandler", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_ObjectiveCMarshalTryGetTaggedMemory", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_OnFirstChanceException", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_OnUnhandledException", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       c_classlibFunctions in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_ProcessFinalizers", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       FinalizerStart(void*) in libRuntime.WorkstationGC.a(FinalizerHelpers.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhRethrow", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhpRethrow in libRuntime.WorkstationGC.a(ExceptionHandling.S.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhThrowEx", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       NotHijacked in libRuntime.WorkstationGC.a(ExceptionHandling.S.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhThrowHwEx", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhpThrowHwEx in libRuntime.WorkstationGC.a(ExceptionHandling.S.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhpCalculateStackTraceWorker", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhpGetCurrentThreadStackTrace in libRuntime.WorkstationGC.a(MiscHelpers.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhpCidResolve", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhpInterfaceDispatchSlow in libRuntime.WorkstationGC.a(StubDispatch.S.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhpFailFastForPInvokeExceptionCoop", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhpPInvokeExceptionGuard in libRuntime.WorkstationGC.a(EHHelpers.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_RhpFailFastForPInvokeExceptionPreemp", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhpPInvokeExceptionGuard in libRuntime.WorkstationGC.a(EHHelpers.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "___managed__Main", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       ___29-[ViewController viewDidLoad]_block_invoke in main.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "___managed__Startup", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       InitializeRuntime() in libbootstrapperdll.o [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_g_compilerEmbeddedKnobsBlob", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       RhConfig::ReadKnobUInt64Value(char const*, unsigned long long*) in libRuntime.WorkstationGC.a(RhConfig.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       RhConfig::ReadKnobBooleanValue(char const*, bool*) in libRuntime.WorkstationGC.a(RhConfig.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       RhConfig::GetKnobNames() in libRuntime.WorkstationGC.a(RhConfig.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       RhConfig::GetKnobValues() in libRuntime.WorkstationGC.a(RhConfig.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       RhConfig::GetKnobCount() in libRuntime.WorkstationGC.a(RhConfig.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_g_compilerEmbeddedSettingsBlob", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       RhConfig::ReadConfigValue(char const*, unsigned long long*, bool) in libRuntime.WorkstationGC.a(RhConfig.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:   "_g_requiredCpuFeatures", referenced from: [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/B02E09DD/p/build/apple/AppleBuild.targets(301,5): error MSB4018:       _RhInitialize in libRuntime.WorkstationGC.a(startup.cpp.o) [/private/tmp/helix/working/B02E09DD/w/AD93097C/e/publish/ProxyProjectForAOTOnHelix.proj]
```
<!-- Known issue validation start -->
 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=367843
**Error message validated:** `Undefined symbols for architecture arm64`
**Result validation: :white_check_mark:** Known issue matched with the provided build.
**Validation performed at:** 8/9/2023 10:18:06 AM UTC
<!-- Known issue validation end -->
<!--Known issue error report start -->
### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[367946](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367946)|dotnet/runtime|[tvOS.Device.Aot.Test.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367946&view=ms.vss-test-web.build-test-results-tab&runId=7792818&resultId=100001)|dotnet/runtime#87773|
|[367843](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367843)|dotnet/runtime|[tvOS.Device.Aot.Test.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367843&view=ms.vss-test-web.build-test-results-tab&runId=7789940&resultId=100001)|dotnet/runtime#87773|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|2|2|
<!--Known issue error report end -->

--- a/src/mono/msbuild/android/build/AndroidBuild.props
+++ b/src/mono/msbuild/android/build/AndroidBuild.props
@@ -8,7 +8,7 @@
     <_HostOS Condition="$([MSBuild]::IsOSPlatform('OSX'))">osx</_HostOS>
     <_HostOS Condition="'$(_HostOS)' == ''">linux</_HostOS>
 
-    <_IsLibraryMode Condition="'$(NativeLib)' != ''">true</_IsLibraryMode>
+    <_IsLibraryMode Condition="'$(UseNativeAOTRuntime)' != 'true' and '$(NativeLib)' != ''">true</_IsLibraryMode>
 
     <AndroidBuildAfterThisTarget Condition="'$(AndroidBuildAfterThisTarget)' == ''">Publish</AndroidBuildAfterThisTarget>
     <AndroidBuildDependsOn>

--- a/src/mono/msbuild/apple/build/AppleBuild.props
+++ b/src/mono/msbuild/apple/build/AppleBuild.props
@@ -19,7 +19,7 @@
     <!-- Tracking issue: https://github.com/dotnet/runtime/issues/87740 -->
     <!-- <ShouldILStrip Condition="'$(RunAOTCompilation)' == 'true' and '$(MonoForceInterpreter)' != 'true'">true</ShouldILStrip> -->
 
-    <_IsLibraryMode Condition="'$(NativeLib)' != ''">true</_IsLibraryMode>
+    <_IsLibraryMode Condition="'$(UseNativeAOTRuntime)' != 'true' and '$(NativeLib)' != ''">true</_IsLibraryMode>
 
     <AppleBuildAfterThisTarget Condition="'$(AppleBuildAfterThisTarget)' == ''">Publish</AppleBuildAfterThisTarget>
     <AppleBuildDependsOn>

--- a/src/tests/build.proj
+++ b/src/tests/build.proj
@@ -414,7 +414,6 @@
     </PropertyGroup>
 
     <PropertyGroup>
-      <NativeDependencies>$(IntermediateOutputPath)\..\$(TestRelativePath)\$(TestName)\native\$(TestName).o</NativeDependencies>
       <DevTeamProvisioning>-</DevTeamProvisioning>
       <AppleGenerateAppBundle>true</AppleGenerateAppBundle>
       <GenerateXcodeProject>true</GenerateXcodeProject>
@@ -422,6 +421,10 @@
       <RunAOTCompilation>false</RunAOTCompilation>
       <MonoForceInterpreter>false</MonoForceInterpreter>
     </PropertyGroup>
+
+    <ItemGroup>
+      <NativeDependencies Include="$(IntermediateOutputPath)\..\$(TestRelativePath)\$(TestName)\native\$(TestName).o" />
+    </ItemGroup>
 
     <ItemGroup>
       <_LinkerFlagsToDrop Include="@(NativeFramework->'-framework %(Identity)')" />

